File: ios/GradeThread/Consignment/ConsignmentReport.swift
import Foundation
import GradeThreadCore

/// One consignor's payout summary over the reported sales (US-676).
struct ConsignmentReportRow: Identifiable, Equatable {
    let consignorId: String
    let consignorName: String
    var itemsSold: Int
    var grossRevenue: Double
    var fees: Double
    /// gross − fees.
    var netProceeds: Double
    /// Amount owed to the consignor (their split of net proceeds).
    var consignorPayout: Double
    /// What the reseller keeps (netProceeds − consignorPayout).
    var yourCut: Double

    var id: String { consignorId }
}

/// Pure per-consignor payout rollup. Net proceeds = sale price − platform fees;
/// the consignor is owed `split%` of that (per-item override, else the
/// consignor's default split); the reseller keeps the remainder. Consignment
/// items are typically acquired at $0 COGS, so COGS is intentionally excluded.
enum ConsignmentReport {

    /// A sold, consigned item flattened to just the numbers the rollup needs —
    /// keeps `compute` independent of SwiftData so it's unit-testable.
    struct SoldConsignedItem: Equatable {
        let consignorId: String
        let splitPctOverride: Double?
        let salePrice: Double
        let fees: Double
    }

    static func compute(
        soldItems: [SoldConsignedItem],
        consignors: [Consignor]
    ) -> [ConsignmentReportRow] {
        let byId = Dictionary(uniqueKeysWithValues: consignors.map { ($0.id, $0) })

        /// One sold item's cents-exact line, so the per-consignor totals can be
        /// summed rather than accumulated.
        struct Line {
            let gross: Double
            let fees: Double
            let net: Double
            let payout: Double
            var yourCut: Double { net - payout }
        }

        let lines: [(consignorId: String, line: Line)] = soldItems.compactMap { item in
            guard let consignor = byId[item.consignorId] else { return nil }
            let split = clampPct(item.splitPctOverride ?? consignor.defaultSplitPct)
            // Round each line item to whole cents (Money.cents): consignorPayout is
            // money OWED to a third party, so `$10.05 × 33.33%` must not carry
            // fractional-cent noise into the payable, and the per-consignor totals
            // must foot against a Money-summed equivalent to the cent.
            let net = Money.cents(item.salePrice - item.fees)
            // Don't pay out on a loss — clamp the consignor's owed amount at 0
            // for a net-negative sale (rare, but a refund-heavy row shouldn't
            // produce a negative "owed").
            let payout = Money.cents(max(net, 0) * split / 100.0)
            return (
                item.consignorId,
                Line(gross: item.salePrice, fees: item.fees, net: net, payout: payout)
            )
        }

        // US-3234: totalled with `Money.sum`, not accumulated with `+=`. The
        // comment above has always said these must foot against a Money-summed
        // equivalent to the cent, and a running `Double` does not — on a
        // consignor with a few hundred sales the payable could differ from the
        // screen's own total, and this is money owed to somebody else.
        return Dictionary(grouping: lines, by: { $0.consignorId })
            .compactMap { consignorId, group -> ConsignmentReportRow? in
                guard let consignor = byId[consignorId] else { return nil }
                let items = group.map(\.line)
                return ConsignmentReportRow(
                    consignorId: consignor.id,
                    consignorName: consignor.name,
                    itemsSold: items.count,
                    grossRevenue: Money.sum(items) { $0.gross },
                    fees: Money.sum(items) { $0.fees },
                    netProceeds: Money.sum(items) { $0.net },
                    consignorPayout: Money.sum(items) { $0.payout },
                    yourCut: Money.sum(items) { $0.yourCut }
                )
            }
            // Most owed first — that's the actionable order (who to pay). Name
            // breaks a tie so the order is stable rather than whatever the
            // dictionary iterated this time.
            .sorted { lhs, rhs in
                if lhs.consignorPayout != rhs.consignorPayout {
                    return lhs.consignorPayout > rhs.consignorPayout
                }
                return lhs.consignorName < rhs.consignorName
            }
    }

    /// Convenience over the local cache: joins sales to their items, keeps only
    /// consigned items, and runs the pure rollup.
    ///
    /// Cancelled/refunded sales are excluded (via `SalePnL.isCompleted`) so a
    /// reversed sale doesn't generate a payout the reseller never collected, and
    /// the fee total mirrors `SalePnL.fees` (platform + payment-processing) so
    /// net proceeds — and therefore the consignor's owed amount — aren't
    /// overstated.
    static func compute(
        items: [LocalInventoryItem],
        sales: [LocalSale],
        consignors: [Consignor]
    ) -> [ConsignmentReportRow] {
        let itemsById = Dictionary(items.map { ($0.id, $0) }) { a, _ in a }
        let sold: [SoldConsignedItem] = sales.compactMap { sale in
            guard SalePnL.isCompleted(sale),
                  let item = itemsById[sale.inventoryItemId],
                  let consignorId = item.consignorId else { return nil }
            return SoldConsignedItem(
                consignorId: consignorId,
                splitPctOverride: item.consignmentSplitPct,
                salePrice: sale.salePrice,
                fees: SalePnL.fees(sale)
            )
        }
        return compute(soldItems: sold, consignors: consignors)
    }

    static func clampPct(_ p: Double) -> Double { min(max(p, 0), 100) }
}

File: android/app/src/main/java/com/gradethread/app/consignment/ConsignmentReport.kt
package com.gradethread.app.consignment

import com.gradethread.app.money.Money
import com.gradethread.app.money.SalePnL
import com.gradethread.app.sync.db.InventoryItemEntity
import com.gradethread.app.sync.db.SaleEntity

/** One consignor's payout summary over the reported sales. */
data class ConsignmentReportRow(
    val consignorId: String,
    val consignorName: String,
    val itemsSold: Int,
    val grossRevenue: Double,
    val fees: Double,
    /** Gross minus fees. */
    val netProceeds: Double,
    /** What the consignor is owed. */
    val consignorPayout: Double,
    /** What the reseller keeps. */
    val yourCut: Double,
)

/**
 * US-1372 (iOS `ConsignmentReport`): who is owed what.
 *
 * Net proceeds are sale price minus fees; the consignor gets their split of
 * that (per-item override first, then the consignor's default); the reseller
 * keeps the rest. Consignment items are typically taken in at zero cost, so
 * COGS is deliberately left out — subtracting a cost the reseller never paid
 * would shrink a payment somebody else is owed.
 *
 * Pure, and rounded per line item, because this is money leaving the business
 * to a third party. `$10.05 × 33.33%` must not carry fractional cents into a
 * payable, and the per-consignor total has to foot to the cent against the
 * lines it came from.
 */
object ConsignmentReport {

    /** A sold, consigned item flattened to just what the rollup needs. */
    data class SoldConsignedItem(
        val consignorId: String,
        val splitPctOverride: Double?,
        val salePrice: Double,
        val fees: Double,
    )

    fun clampPct(value: Double): Double = minOf(maxOf(value, 0.0), 100.0)

    fun compute(
        soldItems: List<SoldConsignedItem>,
        consignors: List<Consignor>,
    ): List<ConsignmentReportRow> {
        val byId = consignors.associateBy { it.id }
        val rows = mutableMapOf<String, ConsignmentReportRow>()

        for (item in soldItems) {
            // An item pointing at a consignor we don't have is skipped rather
            // than reported under a placeholder name — a payout row for
            // "Unknown" is not something anyone can act on.
            val consignor = byId[item.consignorId] ?: continue
            val split = clampPct(item.splitPctOverride ?: consignor.defaultSplitPct)
            val net = Money.cents(item.salePrice - item.fees)
            // No payout on a loss. A refund-heavy line shouldn't produce a
            // negative "owed" that quietly reduces what another sale earned.
            val payout = Money.cents(maxOf(net, 0.0) * split / 100.0)

            val existing = rows[item.consignorId]
            rows[item.consignorId] = ConsignmentReportRow(
                consignorId = consignor.id,
                consignorName = consignor.name,
                itemsSold = (existing?.itemsSold ?: 0) + 1,
                grossRevenue = Money.cents((existing?.grossRevenue ?: 0.0) + item.salePrice),
                fees = Money.cents((existing?.fees ?: 0.0) + item.fees),
                netProceeds = Money.cents((existing?.netProceeds ?: 0.0) + net),
                consignorPayout = Money.cents((existing?.consignorPayout ?: 0.0) + payout),
                yourCut = Money.cents((existing?.yourCut ?: 0.0) + (net - payout)),
            )
        }

        // Most owed first — that is the order someone pays people in.
        return rows.values.sortedWith(
            compareByDescending<ConsignmentReportRow> { it.consignorPayout }
                .thenBy { it.consignorName },
        )
    }

    /**
     * Join sales to their items over the local cache, keep the consigned ones,
     * and run the rollup.
     *
     * Cancelled and refunded sales are excluded through [SalePnL.isCompleted],
     * so a reversed sale can't generate a payout the reseller never collected.
     * Fees mirror [SalePnL.fees] (platform plus payment processing) so net
     * proceeds — and therefore what's owed — aren't overstated.
     */
    fun compute(
        items: List<InventoryItemEntity>,
        sales: List<SaleEntity>,
        consignors: List<Consignor>,
    ): List<ConsignmentReportRow> {
        val itemsById = items.associateBy { it.id }
        val sold = sales.mapNotNull { sale ->
            if (!SalePnL.isCompleted(sale)) return@mapNotNull null
            val item = itemsById[sale.inventoryItemId] ?: return@mapNotNull null
            val consignorId = item.consignorId ?: return@mapNotNull null
            SoldConsignedItem(
                consignorId = consignorId,
                splitPctOverride = item.consignmentSplitPct,
                salePrice = sale.salePrice,
                fees = SalePnL.fees(sale),
            )
        }
        return compute(sold, consignors)
    }

    /** Total owed across every consignor — the number that gets paid out. */
    fun totalOwed(rows: List<ConsignmentReportRow>): Double =
        Money.sum(rows.map { it.consignorPayout })

    fun totalYourCut(rows: List<ConsignmentReportRow>): Double =
        Money.sum(rows.map { it.yourCut })

    /**
     * How many consigned items are still on hand, so the report says something
     * useful before the first sale lands.
     */
    fun unsoldConsignedCount(
        items: List<InventoryItemEntity>,
        sales: List<SaleEntity>,
    ): Int {
        val soldItemIds = sales.filter { SalePnL.isCompleted(it) }
            .map { it.inventoryItemId }
            .toSet()
        return items.count { it.consignorId != null && it.id !in soldItemIds }
    }

    /** The line the report shows when there is nothing to pay yet. */
    fun emptyMessage(consignorCount: Int, unsoldConsigned: Int): String = when {
        consignorCount == 0 ->
            "Add a consignor first, then set one on an item to start tracking what you owe."
        unsoldConsigned == 0 ->
            "No items are assigned to a consignor yet. Set one on an item's page."
        // The distinction matters: items ARE assigned, they just haven't sold.
        // Nothing is wrong and nothing needs doing.
        else ->
            "$unsoldConsigned consigned ${if (unsoldConsigned == 1) "item is" else "items are"} " +
                "waiting to sell. Payouts appear here once they do."
    }
}
